File: lovely-ts/dump/engine/sprite.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__InstanceOf = ____lualib.__TS__InstanceOf
local Error = ____lualib.Error
local RangeError = ____lualib.RangeError
local ReferenceError = ____lualib.ReferenceError
local SyntaxError = ____lualib.SyntaxError
local TypeError = ____lualib.TypeError
local URIError = ____lualib.URIError
local __TS__New = ____lualib.__TS__New
local __TS__Unpack = ____lualib.__TS__Unpack
Sprite = __TS__Class()
Sprite.name = "Sprite"
__TS__ClassExtends(Sprite, Moveable)
function Sprite.prototype.____constructor(self, X, Y, W, H, new_sprite_atlas, sprite_pos)
    Moveable.prototype.____constructor(
        self,
        X,
        Y,
        W,
        H
    )
    self.CT = self.VT
    self.atlas = new_sprite_atlas
    self.scale = {x = self.atlas.px, y = self.atlas.py}
    self.scale_mag = math.min(self.scale.x / W, self.scale.y / H)
    self.zoom = true
    self:set_sprite_pos(sprite_pos)
    if __TS__InstanceOf(self, Sprite) then
        local ____G_I_SPRITE_0 = G.I.SPRITE
        ____G_I_SPRITE_0[#____G_I_SPRITE_0 + 1] = self
    end
end
function Sprite.draw_shader(self, _sprite, arg1)
    error(
        __TS__New(Error, "Method not implemented."),
        0
    )
end
function Sprite.prototype.reset(self)
    self.atlas = G.ASSET_ATLAS[self.atlas.name]
    self:set_sprite_pos(self.sprite_pos)
end
function Sprite.prototype.set_sprite_pos(self, sprite_pos)
    if sprite_pos and sprite_pos.v then
        self.sprite_pos = {
            x = math.random(sprite_pos.v) - 1,
            y = sprite_pos.y
        }
    else
        self.sprite_pos = sprite_pos or ({x = 0, y = 0})
    end
    self.sprite_pos_copy = {x = self.sprite_pos.x, y = self.sprite_pos.y}
    self.sprite = love.graphics.newQuad(
        self.sprite_pos.x * self.atlas.px,
        self.sprite_pos.y * self.atlas.py,
        self.scale.x,
        self.scale.y,
        self.atlas.image:getDimensions()
    )
    self.image_dims = {}
    local ____temp_1 = self.atlas.image:getDimensions()
    self.image_dims[1] = ____temp_1[1]
    self.image_dims[2] = ____temp_1[2]
end
function Sprite.prototype.get_pos_pixel(self)
    self.RETS.get_pos_pixel = self.RETS.get_pos_pixel or ({})
    self.RETS.get_pos_pixel[1] = self.sprite_pos.x
    self.RETS.get_pos_pixel[2] = self.sprite_pos.y
    self.RETS.get_pos_pixel[3] = self.atlas.px
    self.RETS.get_pos_pixel[4] = self.atlas.py
    return self.RETS.get_pos_pixel
end
function Sprite.prototype.get_image_dims(self)
    return self.image_dims
end
function Sprite.prototype.define_draw_steps(self, draw_step_definitions)
    self.draw_steps = EMPTY(_G, self.draw_steps)
    for k, v in ipairs(draw_step_definitions) do
        self.draw_steps[#self.draw_steps + 1 + 1] = {
            shader = v.shader or "dissolve",
            shadow_height = v.shadow_height or nil,
            send = v.send or nil,
            no_tilt = v.no_tilt or nil,
            other_obj = v.other_obj or nil,
            ms = v.ms or nil,
            mr = v.mr or nil,
            mx = v.mx or nil,
            my = v.my or nil
        }
    end
end
function Sprite.prototype.draw_shader(self, _shader, _shadow_height, _send, _no_tilt, other_obj, ms, mr, mx, my, custom_shader, tilt_shadow)
    if G.SETTINGS.reduced_motion then
        _no_tilt = true
    end
    local _draw_major = self.role.draw_major or self
    if _shadow_height then
        self.VT.y = self.VT.y - _draw_major.shadow_parrallax.y * _shadow_height
        self.VT.x = self.VT.x - _draw_major.shadow_parrallax.x * _shadow_height
        self.VT.scale = self.VT.scale * (1 - 0.2 * _shadow_height)
    end
    if custom_shader then
        if _send then
            for k, v in ipairs(_send) do
                G.SHADERS[_shader]:send(
                    v.name,
                    v.val or v.func and v:func() or v.ref_table[v.ref_value]
                )
            end
        end
    end
    if not custom_shader and _shader == "vortex" then
        G.SHADERS.vortex:send("vortex_amt", G.TIMERS.REAL - (G.vortex_time or 0))
    end
    if not custom_shader and _shader ~= "vortex" then
        self.ARGS.prep_shader = self.ARGS.prep_shader or ({})
        self.ARGS.prep_shader.cursor_pos = self.ARGS.prep_shader.cursor_pos or ({})
        self.ARGS.prep_shader.cursor_pos[1] = _draw_major.tilt_var and _draw_major.tilt_var.mx * G.CANV_SCALE or G.CONTROLLER.cursor_position.x * G.CANV_SCALE
        self.ARGS.prep_shader.cursor_pos[2] = _draw_major.tilt_var and _draw_major.tilt_var.my * G.CANV_SCALE or G.CONTROLLER.cursor_position.y * G.CANV_SCALE
        G.SHADERS[_shader or "dissolve"]:send("mouse_screen_pos", self.ARGS.prep_shader.cursor_pos)
        G.SHADERS[_shader or "dissolve"]:send("screen_scale", G.TILESCALE * G.TILESIZE * (_draw_major.mouse_damping or 1) * G.CANV_SCALE)
        G.SHADERS[_shader or "dissolve"]:send("hovering", (_shadow_height and not tilt_shadow or _no_tilt) and 0 or (_draw_major.hover_tilt or 0) * (tilt_shadow or 1))
        G.SHADERS[_shader or "dissolve"]:send(
            "dissolve",
            math.abs(_draw_major.dissolve or 0)
        )
        G.SHADERS[_shader or "dissolve"]:send("time", 123.33412 * (_draw_major.ID / 1.14212 or 12.5123152) % 3000)
        G.SHADERS[_shader or "dissolve"]:send(
            "texture_details",
            self:get_pos_pixel()
        )
        G.SHADERS[_shader or "dissolve"]:send(
            "image_details",
            self:get_image_dims()
        )
        G.SHADERS[_shader or "dissolve"]:send("burn_colour_1", _draw_major.dissolve_colours and _draw_major.dissolve_colours[1] or G.C.CLEAR)
        G.SHADERS[_shader or "dissolve"]:send("burn_colour_2", _draw_major.dissolve_colours and _draw_major.dissolve_colours[2] or G.C.CLEAR)
        G.SHADERS[_shader or "dissolve"]:send("shadow", not not _shadow_height)
        if _send then
            G.SHADERS[_shader or "dissolve"]:send(SMODS.Shaders[_shader or "dissolve"] and SMODS.Shaders[_shader or "dissolve"].original_key or _shader, _send)
        end
    end
    local p_shader = SMODS.Shader.obj_table[_shader or "dissolve"]
    if p_shader and type(p_shader.send_vars) == "function" then
        local sh = G.SHADERS[_shader or "dissolve"]
        local parent_card = self.role.major and self.role.major:is(Card) and self.role.major
        local send_vars = p_shader:send_vars(self, parent_card)
        if type(send_vars) == "table" then
            for key, value in pairs(send_vars) do
                sh:send(key, value)
            end
        end
    end
    love.graphics.setShader(G.SHADERS[_shader or "dissolve"])
    if other_obj then
        self:draw_from(
            other_obj,
            ms,
            mr,
            mx,
            my
        )
    else
        self:draw_this()
    end
    love.graphics.setShader()
    if _shadow_height then
        self.VT.y = self.VT.y + _draw_major.shadow_parrallax.y * _shadow_height
        self.VT.x = self.VT.x + _draw_major.shadow_parrallax.x * _shadow_height
        self.VT.scale = self.VT.scale / (1 - 0.2 * _shadow_height)
    end
end
function Sprite.prototype.draw_this(self, overlay)
    if not self.states.visible then
        return
    end
    if self.sprite_pos.x ~= self.sprite_pos_copy.x or self.sprite_pos.y ~= self.sprite_pos_copy.y then
        self:set_sprite_pos(self.sprite_pos)
    end
    prep_draw(_G, self, 1)
    love.graphics.scale(1 / (self.scale.x / self.VT.w), 1 / (self.scale.y / self.VT.h))
    love.graphics.setColor(__TS__Unpack(overlay or G.BRUTE_OVERLAY or G.C.WHITE))
    if self.video then
        self.video_dims = self.video_dims or ({
            w = self.video:getWidth(),
            h = self.video:getHeight()
        })
        love.graphics.draw(
            self.video,
            0,
            0,
            0,
            self.VT.w / self.T.w / (self.video_dims.w / self.scale.x),
            self.VT.h / self.T.h / (self.video_dims.h / self.scale.y)
        )
    else
        love.graphics.draw(
            self.atlas.image,
            self.sprite,
            0,
            0,
            0,
            self.VT.w / self.T.w,
            self.VT.h / self.T.h
        )
    end
    love.graphics.pop()
    add_to_drawhash(_G, self)
    self:draw_boundingrect()
    if self.shader_tab then
        love.graphics.setShader()
    end
end
function Sprite.prototype.draw(self, overlay)
    if not self.states.visible then
        return
    end
    if self.draw_steps then
        for k, v in ipairs(self.draw_steps) do
            local ____self_draw_shader_4 = self.draw_shader
            local ____v_shader_3 = v.shader
            local ____v_shadow_height_2 = v.shadow_height
            if ____v_shadow_height_2 == nil then
                ____v_shadow_height_2 = 0
            end
            ____self_draw_shader_4(
                self,
                ____v_shader_3,
                ____v_shadow_height_2,
                v.send,
                v.no_tilt,
                v.other_obj,
                v.ms,
                v.mr,
                v.mx,
                v.my,
                not not v.send
            )
        end
    else
        self:draw_this(overlay)
    end
    add_to_drawhash(_G, self)
    for k, v in pairs(self.children) do
        if k ~= "h_popup" then
            v:draw()
        end
    end
    add_to_drawhash(_G, self)
    self:draw_boundingrect()
end
function Sprite.prototype.draw_from(self, other_obj, ms, mr, mx, my)
    self.ARGS.draw_from_offset = self.ARGS.draw_from_offset or ({})
    self.ARGS.draw_from_offset.x = mx or 0
    self.ARGS.draw_from_offset.y = my or 0
    prep_draw(
        _G,
        other_obj,
        1 + (ms or 0),
        mr or 0,
        self.ARGS.draw_from_offset,
        true
    )
    love.graphics.scale(1 / (other_obj.scale_mag or other_obj.VT.scale))
    love.graphics.setColor(G.BRUTE_OVERLAY or G.C.WHITE)
    love.graphics.draw(
        self.atlas.image,
        self.sprite,
        -(other_obj.T.w / 2 - other_obj.VT.w / 2) * 10,
        0,
        0,
        other_obj.VT.w / other_obj.T.w,
        other_obj.VT.h / other_obj.T.h
    )
    self:draw_boundingrect()
    love.graphics.pop()
end
function Sprite.prototype.remove(self)
    if self.video then
        self.video:release()
    end
    for k, v in pairs(G.ANIMATIONS) do
        if v == self then
            table.remove(G.ANIMATIONS, k)
        end
    end
    for k, v in pairs(G.I.SPRITE) do
        if v == self then
            table.remove(G.I.SPRITE, k)
        end
    end
    Moveable.prototype.remove(self)
end

File: lovely-ts/dump/engine/sprite.ts
///<reference types="lua-types/jit"/>
///<reference types="love-typescript-definitions"/>
///<reference path="node.ts"/>
///<reference path="moveable.ts"/>
class Sprite extends Moveable {
    atlas: any;
    scale: Position2D;
    scale_mag: number;
    sprite_pos_copy: Position2D;
    sprite: import("love.graphics").Quad;
    image_dims: {};
    draw_steps: DrawStepDefinition[];
    video: any;
    video_dims: any;
    shader_tab: any;
    sprite_pos: {x:number;y:number;v?:number};
    constructor(X: number, Y: number, W: number, H: number, new_sprite_atlas: any, sprite_pos: Position2D) {
        super(X,Y,W,H);
        this.CT = this.VT;
        this.atlas = new_sprite_atlas;
        this.scale = { x: this.atlas.px, y: this.atlas.py };
        this.scale_mag = Math.min(this.scale.x / W, this.scale.y / H);
        this.zoom = true;
        this.set_sprite_pos(sprite_pos);
        if (this instanceof Sprite) {
            G.I.SPRITE.push(this);
        }
    }
    reset() {
        this.atlas = G.ASSET_ATLAS[this.atlas.name];
        this.set_sprite_pos(this.sprite_pos);
    };
    set_sprite_pos(sprite_pos: { v?: number; y: number; x: number; }) {
        if (sprite_pos && sprite_pos.v) {
            this.sprite_pos = { x: math.random(sprite_pos.v) - 1, y: sprite_pos.y };
        }
        else {
            this.sprite_pos = sprite_pos || { x: 0, y: 0 };
        }
        this.sprite_pos_copy = { x: this.sprite_pos.x, y: this.sprite_pos.y };
        this.sprite = love.graphics.newQuad(this.sprite_pos.x * this.atlas.px, this.sprite_pos.y * this.atlas.py, this.scale.x, this.scale.y, this.atlas.image.getDimensions());
        this.image_dims = {};
        [this.image_dims[1], this.image_dims[2]] = this.atlas.image.getDimensions();
    };
    get_pos_pixel() {
        this.RETS.get_pos_pixel = this.RETS.get_pos_pixel ?? [];
        this.RETS.get_pos_pixel[0] = this.sprite_pos.x;
        this.RETS.get_pos_pixel[1] = this.sprite_pos.y;
        this.RETS.get_pos_pixel[2] = this.atlas.px;
        this.RETS.get_pos_pixel[3] = this.atlas.py;
        return this.RETS.get_pos_pixel;
    };
    get_image_dims() {
        return this.image_dims;
    };
    define_draw_steps(draw_step_definitions: DrawStepDefinition[]) {
        this.draw_steps = EMPTY(this.draw_steps);
        for (const [k, v] of ipairs(draw_step_definitions)) {
            this.draw_steps[this.draw_steps.length + 1] = { shader: v.shader || "dissolve", shadow_height: v.shadow_height || undefined, send: v.send || undefined, no_tilt: v.no_tilt || undefined, other_obj: v.other_obj || undefined, ms: v.ms || undefined, mr: v.mr || undefined, mx: v.mx || undefined, my: v.my || undefined };
        }
    };
    draw_shader(_shader: string, _shadow_height: number, _send?: DrawStepSend[], _no_tilt?: boolean, other_obj?: any, ms?: number, mr?: number, mx?: number, my?: number, custom_shader?: boolean, tilt_shadow?: number) {
        if (G.SETTINGS.reduced_motion) {
            _no_tilt = true;
        }
        let _draw_major = this.role.draw_major || this;
        if (_shadow_height) {
            this.VT.y = this.VT.y - _draw_major.shadow_parrallax.y * _shadow_height;
            this.VT.x = this.VT.x - _draw_major.shadow_parrallax.x * _shadow_height;
            this.VT.scale = this.VT.scale * (1 - 0.2 * _shadow_height);
        }
        if (custom_shader) {
            if (_send) {
                for (const [k, v] of ipairs(_send)) {
                    G.SHADERS[_shader].send(v.name, v.val || v.func && v.func() || v.ref_table[v.ref_value]);
                }
            }
        }
        if (!custom_shader && _shader === "vortex") {
            G.SHADERS["vortex"].send("vortex_amt", G.TIMERS.REAL - (G.vortex_time || 0));
        }
        if (!custom_shader && _shader !== "vortex") {
            this.ARGS.prep_shader = this.ARGS.prep_shader || {};
            this.ARGS.prep_shader.cursor_pos = this.ARGS.prep_shader.cursor_pos || {};
            this.ARGS.prep_shader.cursor_pos[1] = _draw_major.tilt_var && _draw_major.tilt_var.mx * G.CANV_SCALE || G.CONTROLLER.cursor_position.x * G.CANV_SCALE;
            this.ARGS.prep_shader.cursor_pos[2] = _draw_major.tilt_var && _draw_major.tilt_var.my * G.CANV_SCALE || G.CONTROLLER.cursor_position.y * G.CANV_SCALE;
            G.SHADERS[_shader || "dissolve"].send("mouse_screen_pos", this.ARGS.prep_shader.cursor_pos);
            G.SHADERS[_shader || "dissolve"].send("screen_scale", G.TILESCALE * G.TILESIZE * (_draw_major.mouse_damping || 1) * G.CANV_SCALE);
            G.SHADERS[_shader || "dissolve"].send("hovering", (_shadow_height && !tilt_shadow || _no_tilt) && 0 || (_draw_major.hover_tilt || 0) * (tilt_shadow || 1));
            G.SHADERS[_shader || "dissolve"].send("dissolve", Math.abs(_draw_major.dissolve || 0));
            G.SHADERS[_shader || "dissolve"].send("time", 123.33412 * (_draw_major.ID / 1.14212 || 12.5123152) % 3000);
            G.SHADERS[_shader || "dissolve"].send("texture_details", this.get_pos_pixel());
            G.SHADERS[_shader || "dissolve"].send("image_details", this.get_image_dims());
            G.SHADERS[_shader || "dissolve"].send("burn_colour_1", _draw_major.dissolve_colours && _draw_major.dissolve_colours[1] || G.C.CLEAR);
            G.SHADERS[_shader || "dissolve"].send("burn_colour_2", _draw_major.dissolve_colours && _draw_major.dissolve_colours[2] || G.C.CLEAR);
            G.SHADERS[_shader || "dissolve"].send("shadow", !!_shadow_height);
            if (_send) {
                G.SHADERS[_shader || "dissolve"].send(SMODS.Shaders[_shader || "dissolve"] && SMODS.Shaders[_shader || "dissolve"].original_key || _shader, _send);
            }
        }
        let p_shader = SMODS.Shader.obj_table[_shader || "dissolve"];
        if (p_shader && type(p_shader.send_vars) === "function") {
            let sh = G.SHADERS[_shader || "dissolve"];
            let parent_card = this.role.major && this.role.major.is(Card) && this.role.major;
            let send_vars = p_shader.send_vars(this, parent_card);
            if (type(send_vars) === "table") {
                for (const [key, value] of pairs(send_vars)) {
                    sh.send(key as (string|number), value);
                }
            }
        }
        love.graphics.setShader(G.SHADERS[_shader || "dissolve"], /**G.SHADERS[_shader || "dissolve"]*/);
        if (other_obj) {
            this.draw_from(other_obj, ms, mr, mx, my);
        }
        else {
            this.draw_this();
        }
        love.graphics.setShader();
        if (_shadow_height) {
            this.VT.y = this.VT.y + _draw_major.shadow_parrallax.y * _shadow_height;
            this.VT.x = this.VT.x + _draw_major.shadow_parrallax.x * _shadow_height;
            this.VT.scale = this.VT.scale / (1 - 0.2 * _shadow_height);
        }
    };
    draw_this(overlay?: HexArray) {
        if (!this.states.visible) {
            return;
        }
        if (this.sprite_pos.x !== this.sprite_pos_copy.x || this.sprite_pos.y !== this.sprite_pos_copy.y) {
            this.set_sprite_pos(this.sprite_pos);
        }
        prep_draw(this, 1);
        love.graphics.scale(1 / (this.scale.x / this.VT.w), 1 / (this.scale.y / this.VT.h));
        love.graphics.setColor(...(overlay || G.BRUTE_OVERLAY || G.C.WHITE));
        if (this.video) {
            this.video_dims = this.video_dims || { w: this.video.getWidth(), h: this.video.getHeight() };
            love.graphics.draw(this.video, 0, 0, 0, this.VT.w / this.T.w / (this.video_dims.w / this.scale.x), this.VT.h / this.T.h / (this.video_dims.h / this.scale.y));
        }
        else {
            love.graphics.draw(this.atlas.image, this.sprite, 0, 0, 0, this.VT.w / this.T.w, this.VT.h / this.T.h);
        }
        love.graphics.pop();
        add_to_drawhash(this);
        this.draw_boundingrect();
        if (this.shader_tab) {
            love.graphics.setShader();
        }
    };
    draw(overlay?: any) {
        if (!this.states.visible) {
            return;
        }
        if (this.draw_steps) {
            for (const [k, v] of ipairs(this.draw_steps)) {
                this.draw_shader(v.shader, v.shadow_height??0, v.send, v.no_tilt, v.other_obj, v.ms, v.mr, v.mx, v.my, !!v.send);
            }
        }
        else {
            this.draw_this(overlay);
        }
        add_to_drawhash(this);
        for (const [k, v] of pairs(this.children)) {
            if (k !== "h_popup") {
                v.draw();
            }
        }
        add_to_drawhash(this);
        this.draw_boundingrect();
    };
    draw_from(other_obj: Sprite, ms?: number, mr?: number, mx?: number, my?: number) {
        this.ARGS.draw_from_offset = this.ARGS.draw_from_offset || {};
        this.ARGS.draw_from_offset.x = mx || 0;
        this.ARGS.draw_from_offset.y = my || 0;
        prep_draw(other_obj, 1 + (ms || 0), mr || 0, this.ARGS.draw_from_offset, true);
        love.graphics.scale(1 / (other_obj.scale_mag || other_obj.VT.scale));
        love.graphics.setColor(G.BRUTE_OVERLAY || G.C.WHITE);
        love.graphics.draw(this.atlas.image, this.sprite, -(other_obj.T.w / 2 - other_obj.VT.w / 2) * 10, 0, 0, other_obj.VT.w / other_obj.T.w, other_obj.VT.h / other_obj.T.h);
        this.draw_boundingrect();
        love.graphics.pop();
    };
    remove() {
        if (this.video) {
            this.video.release();
        }
        for (const [k, v] of pairs(G.ANIMATIONS)) {
            if (v === this) {
                table.remove(G.ANIMATIONS, k as (number|undefined));
            }
        }
        for (const [k, v] of pairs(G.I.SPRITE)) {
            if (v === this) {
                table.remove(G.I.SPRITE, k as (number|undefined));
            }
        }
        Moveable.prototype.remove.call(this);
    };
}